locations/middlewares/anti_bot_detection.py

```diff
@@ -0,0 +1,103 @@
+    # https://learn.microsoft.com/en-us/azure/web-application-firewall/ag/bot-protection-overview
+    AZURE_WAF = {"name": "Azure WAF", "zyte_bypassable": True}
+
+    # Cloudflare bot protection documentation:
+    # https://developers.cloudflare.com/bots/
+    CLOUDFLARE = {"name": "Cloudflare", "zyte_bypassable": True}
+
+    # DataDome bot protection product brochure:
+    # https://datadome.co/products/bot-protection/
+    DATADOME = {"name": "DataDome", "zyte_bypassable": True}
+
+    # HUMAN Bot Defender bot protection product brochure:
+    # https://www.humansecurity.com/products/human-bot-defender
+    HUMAN = {"name": "HUMAN", "zyte_bypassable": True}
+
+    # Imperva bot protection product brochure:
+    # https://www.imperva.com/products/bot-detection-mitigation/
+    IMPERVA = {"name": "Imperva", "zyte_bypassable": True}
+
+    # Qrator bot protection documentation:
+    # https://docs.qrator.net/en/technologies/tracking-cookie.html
+    QRATOR = {"name": "Qrator", "zyte_bypassable": True}
+
+
+class AntiBotDetectionMiddleware:
+    def __init__(self, crawler):
+        self.crawler = crawler
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def process_response(self, request: Request, response: Response, spider: Spider):
+        if response.headers.get("cf-mitigated"):
+            self.add_anti_bot_method(AntiBotMethods.CLOUDFLARE, spider)
+
+        if response.headers.get("x-datadome"):
+            self.add_anti_bot_method(AntiBotMethods.DATADOME, spider)
+
+        if server := response.headers.get("server"):
+            server_utf8 = self.decode_http_header_value(server)
+            if server_utf8.upper() == "QRATOR":
+                self.add_anti_bot_method(AntiBotMethods.QRATOR, spider)
+            if response.status == 403 and server_utf8.upper() == "MICROSOFT-AZURE-APPLICATION-GATEWAY/V2":
+                self.add_anti_bot_method(AntiBotMethods.AZURE_WAF, spider)
+
+        if cookies := response.headers.getlist("set-cookie"):
+            for cookie in cookies:
+                cookie_utf8 = self.decode_http_header_value(cookie)
+                if cookie_utf8.startswith("visid_incap_") or cookie_utf8.startswith("incap_ses_"):
+                    self.add_anti_bot_method(AntiBotMethods.IMPERVA, spider)
+
+        if response.status == 403:
+            # We may need to decompress the response body before it
+            # can be parsed to detect anti-bot methods. This is
+            # necessary because HttpErrorMiddleware is ordered
+            # before HttpCompressionMiddleware per:
+            # 1. https://docs.scrapy.org/en/latest/topics/settings.html#spider-middlewares-base
+            # 2. https://docs.scrapy.org/en/latest/topics/settings.html#downloader-middlewares-base
+            # It is probably not a good idea to reorder these
+            # default middleware orders.
+            httpcm = HttpCompressionMiddleware.from_crawler(spider.crawler)
```

Comment:
This probably shouldn't be initiated on emery request

Reply:
This only triggers on a HTTP 403 error though, which would typically result in a spider terminating its crawl. I don't think we'll see a performance impact as a result, because we wouldn't expect more than a few HTTP 403 errors (>1 because multiple requests could have been fired at once, and perhaps 10 HTTP 403 errors are still being sent back to Scrapy for handling).

AFAIK Azure WAF sometimes only triggers after X number of requests in Y seconds. So the first 20 requests could be fine, then request 21 triggers a HTTP 403 error and CAPTCHA challenge from Azure WAF.